File: ios/Runner/ImageProcessingChannel.swift
import Flutter
import Foundation

/// Thrown for argument problems detected on the Swift side before ever reaching OpenCV.
private struct ArgumentError: Error {
    let message: String
}

/// Bridges Dart to `ImageProcessingOpenCV` (Obj-C++/OpenCV) via a single MethodChannel. Every
/// call is dispatched onto a background queue — never the main thread — since these are
/// multi-hundred-ms to multi-second operations on full-resolution scan images.
enum ImageProcessingChannel {
    static let channelName = "beacon_smart_scan/image_processing"

    static func register(messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
        channel.setMethodCallHandler { call, result in
            DispatchQueue.global(qos: .userInitiated).async {
                let response = handle(call: call)
                DispatchQueue.main.async { result(response) }
            }
        }
    }

    private static func handle(call: FlutterMethodCall) -> Any {
        guard let args = call.arguments as? [String: Any] else {
            return errorResult(code: "INVALID_ARGUMENT", message: "Missing arguments")
        }
        do {
            switch call.method {
            case "sharpen":
                let inputPath = try requireString(args, "inputPath")
                let outputPath = try requireString(args, "outputPath")
                let amount = (args["amount"] as? Double) ?? 1.5
                let radius = (args["radius"] as? Double) ?? 3.0
                try ImageProcessingOpenCV.sharpen(atPath: inputPath, outputPath: outputPath, amount: amount, radius: radius)
                return ["outputPath": outputPath]

            case "removeShadow":
                let inputPath = try requireString(args, "inputPath")
                let outputPath = try requireString(args, "outputPath")
                try ImageProcessingOpenCV.removeShadow(atPath: inputPath, outputPath: outputPath)
                return ["outputPath": outputPath]

            case "rotate":
                let inputPath = try requireString(args, "inputPath")
                let outputPath = try requireString(args, "outputPath")
                let quarterTurns = (args["quarterTurnsClockwise"] as? Int) ?? 1
                try ImageProcessingOpenCV.rotate(atPath: inputPath, outputPath: outputPath, quarterTurnsClockwise: quarterTurns)
                return ["outputPath": outputPath]

            case "detectHandwritingRegions":
                let imagePath = try requireString(args, "imagePath")
                let textBlocks = (args["textBlocks"] as? [[String: Any]]) ?? []
                return try ImageProcessingOpenCV.detectHandwritingRegions(atPath: imagePath, textBlocks: textBlocks)

            case "classifyHandwriting":
                let imagePath = try requireString(args, "imagePath")
                let textBlocks = (args["textBlocks"] as? [[String: Any]]) ?? []
                return try MlHandwritingClassifier.classify(imagePath: imagePath, textBlocks: textBlocks)

            case "eraseRegions":
                let inputPath = try requireString(args, "inputPath")
                let outputPath = try requireString(args, "outputPath")
                let rects = (args["rects"] as? [[String: Any]]) ?? []
                let padding = (args["padding"] as? Double) ?? 6.0
                let inpaintRadius = (args["inpaintRadius"] as? Double) ?? 5.0
                try ImageProcessingOpenCV.eraseRegions(atPath: inputPath, outputPath: outputPath, rects: rects, padding: padding, inpaintRadius: inpaintRadius)
                return ["outputPath": outputPath]

            default:
                return errorResult(code: "INVALID_ARGUMENT", message: "Unknown method: \(call.method)")
            }
        } catch let error as ArgumentError {
            return errorResult(code: "INVALID_ARGUMENT", message: error.message)
        } catch let error as NSError where error.domain == ImageProcessingErrorDomain {
            return errorResult(code: code(forOpenCvError: error), message: error.localizedDescription)
        } catch {
            return errorResult(code: "PROCESSING_FAILED", message: error.localizedDescription)
        }
    }

    private static func requireString(_ args: [String: Any], _ key: String) throws -> String {
        guard let value = args[key] as? String else {
            throw ArgumentError(message: "Missing argument: \(key)")
        }
        return value
    }

    private static func code(forOpenCvError error: NSError) -> String {
        switch ImageProcessingErrorCode(rawValue: error.code) {
        case .fileNotFound: return "FILE_NOT_FOUND"
        case .invalidArgument: return "INVALID_ARGUMENT"
        default: return "PROCESSING_FAILED"
        }
    }

    private static func errorResult(code: String, message: String) -> FlutterError {
        FlutterError(code: code, message: message, details: nil)
    }
}

File: android/app/src/main/kotlin/com/example/beacon_smart_scan/imageprocessing/ImageProcessingChannel.kt
package com.example.beacon_smart_scan.imageprocessing

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executors
import org.opencv.android.OpenCVLoader

/** Errors thrown by the native image-processing pipeline, mapped 1:1 to a [MethodChannel.Result.error] code. */
class ImageProcessingException(val code: String, message: String) : Exception(message)

/**
 * Bridges Dart to the native OpenCV pipeline via a single MethodChannel.
 * Every call is dispatched onto [executor] (never the platform/UI thread) since these are
 * multi-hundred-ms to multi-second operations on full-resolution scan images.
 */
object ImageProcessingChannel {
    private const val CHANNEL_NAME = "beacon_smart_scan/image_processing"
    private val executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())
    @Volatile private var openCvLoaded = false
    private lateinit var appContext: Context

    fun register(flutterEngine: FlutterEngine, context: Context) {
        appContext = context.applicationContext
        val channel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler { call, result ->
            executor.execute {
                try {
                    ensureOpenCvLoaded()
                    val response = handle(call)
                    mainHandler.post { result.success(response) }
                } catch (e: ImageProcessingException) {
                    mainHandler.post { result.error(e.code, e.message, null) }
                } catch (e: Exception) {
                    mainHandler.post { result.error("PROCESSING_FAILED", e.message, null) }
                }
            }
        }
    }

    private fun ensureOpenCvLoaded() {
        if (openCvLoaded) return
        if (!OpenCVLoader.initLocal()) {
            throw ImageProcessingException("PROCESSING_FAILED", "Không thể khởi tạo thư viện OpenCV")
        }
        openCvLoaded = true
    }

    private fun handle(call: MethodCall): Any {
        return when (call.method) {
            "sharpen" ->
                Sharpener.sharpen(
                    requirePath(call, "inputPath"),
                    requirePath(call, "outputPath"),
                    requireDouble(call, "amount", 1.5),
                    requireDouble(call, "radius", 3.0),
                )
            "removeShadow" ->
                ShadowRemover.removeShadow(
                    requirePath(call, "inputPath"),
                    requirePath(call, "outputPath"),
                )
            "rotate" ->
                Rotator.rotate(
                    requirePath(call, "inputPath"),
                    requirePath(call, "outputPath"),
                    (call.argument<Any>("quarterTurnsClockwise") as? Number)?.toInt() ?: 1,
                )
            "detectOrphanRegions" ->
                OrphanInkDetector.detect(
                    requirePath(call, "imagePath"),
                    requireMapList(call, "existingBlocks"),
                )
            "detectHandwritingRegions" ->
                HandwritingDetector.detect(
                    requirePath(call, "imagePath"),
                    requireMapList(call, "textBlocks"),
                )
            "classifyHandwriting" ->
                MlHandwritingClassifier.classify(
                    appContext,
                    requirePath(call, "imagePath"),
                    requireMapList(call, "textBlocks"),
                )
            "eraseRegions" ->
                Inpainter.erase(
                    requirePath(call, "inputPath"),
                    requirePath(call, "outputPath"),
                    requireMapList(call, "rects"),
                    requireMapList(call, "keepRects"),
                    requireDouble(call, "padding", 6.0),
                    requireDouble(call, "inpaintRadius", 5.0),
                )
            else -> throw ImageProcessingException("INVALID_ARGUMENT", "Unknown method: ${call.method}")
        }
    }

    private fun requirePath(call: MethodCall, key: String): String =
        call.argument<String>(key)
            ?: throw ImageProcessingException("INVALID_ARGUMENT", "Missing argument: $key")

    private fun requireDouble(call: MethodCall, key: String, default: Double): Double =
        (call.argument<Any>(key) as? Number)?.toDouble() ?: default

    @Suppress("UNCHECKED_CAST")
    private fun requireMapList(call: MethodCall, key: String): List<Map<String, Any>> =
        (call.argument<List<Any>>(key) ?: emptyList()) as List<Map<String, Any>>
}
